### ISSUE TITLE
Export from "app" must be a function that returns an HTML string

### ISSUE THREAD

**Issue by sidjain26** [markdalgleish/static-site-generator-webpack-plugin#46]
Here's a repo that reproduces the issue: [https://github.com/sidjain26/static-react-issue](https://github.com/sidjain26/static-react-issue)

Note: This is using `webpack v2.1`, `babel v6`, `react v15`, `react-router v2.4`

Run 
```
npm i
npm run build
```

**Comment by stevenpetryk** [comment]
Encountering the same problem. It seems to only happen when exporting as CommonJS.

**Comment by stevenpetryk** [comment]
@sidjain26, I realized that the `paths` being passed into the plugin didn't match my `react-router` routes. Changing this fixed my issue. Maybe give it a look?

**Comment by sidjain26** [comment]
I am not sure if this is correct but I am passing the folloeing paths:

```javascript
const data = {
  routes: [
    '/',
    '/about'
  ]
};
```
The routes looks like:

```javascript

const routes = (
  <Route path="/" component={App}>
    <IndexRoute component={Home} />
    <Route path="/about" component={About} />
  </Route>
);
```

**Comment by stevenpetryk** [comment]
Inside of your callback passed to react-router's `match()`, assuming that your main method looks like this:

```js
export default (locals, callback) => {
  const history = createMemoryHistory()
  const location = history.createLocation(locals.path)

  match({ routes, location }, (error, redirectLocation, renderProps) => {
    const app = ReactDOMServer.renderToString(<RouterContext {...renderProps} />)

    callback(null, baseHtml({ app}))
  })
}
```

Could you log `renderProps` and post what gets logged? It'll log twice since you have two routes.

**Comment by sidjain26** [comment]
What is `baseHtml` in the main method above? `renderToString` outputs html ny itself

**Comment by stevenpetryk** [comment]
@sidjain26, it's a simple method that renders a template string. It's just for my HTML boilerplate.

Something like this:

```js
// baseHtml.js
export default ({ app }) => `
  <html>
    <head>
      ...
    </head>

    <body>
      <div id="app">${app}</div>
    </body>
  </html>
`
```

**Comment by sidjain26** [comment]
For the purpose of reproducing the issue, I guess this is redundant. Passing the output of `renderToString` to `callback` should work. Correct me if I am wrong.

**Comment by d4goxn** [comment]
Does anyone have a complete working example of static site generation with this module? There is a fairly complete blog post at http://www.qimingweng.com/writing/webpack-static-render, but the signature of the function that generates the HTML differs from that of the [readme](https://github.com/markdalgleish/static-site-generator-webpack-plugin#react-router-example). Both that blog post and the instructions in the readme lead to this same issue for me:

```
ERROR in Error: Export from "main" must be a function that returns an HTML string
```

**Comment by ahoereth** [comment]
Problem seems to be that the `render` variable is empty after running `evaluate` on the source. I tried replacing it with node's [`runInThisContext`](https://nodejs.org/api/vm.html#vm_vm_runinthiscontext_code_options) -- just returns false. The source code passed into the `evaluate` method looks fine tho..

**Comment by guilhermebruzzi** [comment]
"Export from "app" must be a function that returns an HTML string" is an error indicating that something is wrong in your index.(js|jsx) file.
To debug this kind of error, i copied this plugin to a local path and inserted a console.log on the error in the catch block and also a try catch in my index.js file. In my case it was a problem that was solved using the global scope parameter.
I inserted a console.log in the plugin's catch block, because  in my case and for some reason the `compiler.errors.push(err.stack);` statement of this plugin wasn't printing the error on the console, but when i tried to run @sidjain26 example, it did printed.
The error in his example was a babel error: "Cannot remove 'babel-plugin-transform-es2015-modules-commonjs' from the plugin list.".
I submited a PR to his example https://github.com/sidjain26/static-react-issue/pull/1 (just needed to replace "babel-preset-es2015-webpack" to "babel-preset-es2015" in .babelrc and package.json).

**Comment by ahoereth** [comment]
I am using

    "presets": [
      "react",
      ["es2015", { "modules": false } ]
    ]

and setting [`output.libraryTarget: 'umd'`](https://webpack.github.io/docs/configuration.html#output-librarytarget) in my webpack config fixed it for me. Still.. there should be a way to make this work with es6 modules.

**Comment by guilhermebruzzi** [comment]
@ahoereth I does work with es6 modules, see the https://github.com/guilhermebruzzi/static-react-issue example that i forked from @sidjain26 
.babelrc:
{
  "presets": ["babel-preset-es2015", "react"],
  "plugins": ["transform-class-properties", "transform-runtime"]
}

**Comment by ahoereth** [comment]
I saw your comment and pull request. In order to leverage tree shaking one needs to disable module syntax transformation.

> webpack 2 natively supports ES6 import and export statements. webpack 2 leverages the static structure of the ES6 modules to perform tree shaking.

More info: https://github.com/gajus/babel-preset-es2015-webpack

**Comment by sidjain26** [comment]
Any progress on this anyone?

**Comment by sidjain26** [comment]
@ahoereth that preset has been deprecated. See https://github.com/gajus/babel-preset-es2015-webpack#deprecated

**Comment by ahoereth** [comment]
@sidjain26 yes, please read the link you posted:

> As of Babel version v6.13.0, babel-preset-es2015 supports modules option.
> Use **preset options to disable transformation of the ES6 module syntax**:
```json
{
    "presets": [
        [
            "es2015",
            {
                "modules": false
            }
        ]
    ]
}
```

**Comment by sidjain26** [comment]
@ahoereth lol you misunderstood, I meant we can use the original es2015 preset instead of the webpack version now.

Also, I had some time to look into this issue and I found the error. The problem seems to be with using ES2015 modules without transpiling them with Babel. This prevents https://github.com/markdalgleish/static-site-generator-webpack-plugin/blob/997318a3fccc765dfdf77bed6891bd1be20b0ed5/index.js#L33-L35 from executing. Setting `render` to `render["default"]` makes the plugin work as expected without any errors. I will submit a PR soon.

**Comment by sidjain26** [comment]
@ahoereth lol you misunderstood, I meant we can use the original es2015 preset instead of the webpack version now.

Also, I had some time to look into this issue and I found the error. The problem seems to be with using ES2015 modules without transpiling them babel. This

**Comment by ahoereth** [comment]
@sidjain26 which is exactly what I described in [my comment above](https://github.com/markdalgleish/static-site-generator-webpack-plugin/issues/46#issuecomment-240685332) :) Thank you for your PR! It should maybe include a check for both: Either `default` or `__esModule` should be an own property?

**Comment by sidjain26** [comment]
@ahoereth Oh nice, I missed your comment while going through this. 
Also, that is exactly what I did in the original PR but I modified it to just check for default, since if `__esModule` is own property then, so will be `default`. So didn't make sense to check for both.

**Comment by markdalgleish** [comment]
Fixed in `v2.1.1` 👍

**Comment by sidjain26** [comment]
It would have been better to do `export const __esModule = true` in the main entry in hindsight. This is a webpack issue.
https://github.com/webpack/webpack/issues/2945#issuecomment-244733810
